lib/python/qmk/flashers.py

```diff
@@ -0,0 +1,126 @@
+import shutil
+import time
+
+import usb.core
+
+from qmk.constants import BOOTLOADER_VIDS_PIDS
+from milc import cli
+
+# yapf: disable
+_PID_TO_MCU = {
+    '2fef': 'atmega16u2',
+    '2ff0': 'atmega32u2',
+    '2ff3': 'atmega16u4',
+    '2ff4': 'atmega32u4',
+    '2ff9': 'at90usb64',
+    '2ffa': 'at90usb162',
+    '2ffb': 'at90usb128'
+}
+
+AVRDUDE_MCU = {
+    'atmega32a': 'm32',
+    'atmega328p': 'm328p',
+    'atmega328': 'm328',
+}
+# yapf: enable
+
+
+def _find_bootloader():
+    # To avoid running forever in the background, only look for bootloaders for 10min
+    start_time = time.time()
+    while time.time() - start_time < 600:
+        for bl in BOOTLOADER_VIDS_PIDS:
+            for vid, pid in BOOTLOADER_VIDS_PIDS[bl]:
+                vid_hex = int(f'0x{vid}', 0)
+                pid_hex = int(f'0x{pid}', 0)
+                dev = usb.core.find(idVendor=vid_hex, idProduct=pid_hex)
+                if dev:
+                    if bl == 'atmel-dfu':
+                        details = _PID_TO_MCU[pid]
+                    elif bl == 'caterina':
+                        details = (vid_hex, pid_hex)
+                    elif bl == 'hid-bootloader':
+                        if vid == '16c0' and pid == '0478':
+                            details = 'halfkay'
+                        else:
+                            details = 'qmk-hid'
+                    elif bl == 'stm32' or bl == 'apm32':
+                        details = (vid, pid)
+                    else:
+                        details = None
+                    return (bl, details)
+    return (None, None)
+
+
+def _find_serial_port(vid, pid):
+    if 'windows' in cli.platform.lower():
+        from serial.tools.list_ports_windows import comports
+    else:
+        from serial.tools.list_ports_posix import comports
+    for port in comports():
+        port, desc, hwid = port
+        if f'{vid:04x}:{pid:04x}' in hwid.casefold():
+            return port
+    return None
+
+
+def _flash_caterina(details, file):
+    port = _find_serial_port(details[0], details[1])
+    cli.run(['avrdude', '-p', 'atmega32u4', '-c', 'avr109', '-U', f'flash:w:{file}:i', '-P', port], capture_output=False)
+
+
+def _flash_atmel_dfu(mcu, file):
+    cli.run(['dfu-programmer', mcu, 'erase', '--force'], capture_output=False)
+    cli.run(['dfu-programmer', mcu, 'flash', '--force', file], capture_output=False)
+    cli.run(['dfu-programmer', mcu, 'reset'], capture_output=False)
+
+
+def _flash_hid_bootloader(mcu, details, file):
+    if details == 'halfkay':
+        if shutil.which('teensy-loader-cli'):
+            cmd = 'teensy-loader-cli'
+        elif shutil.which('teensy_loader_cli'):
+            cmd = 'teensy_loader_cli'
+    else:
```

Comment:
Apparently `hid_bootloader_cli` should be able to flash the `halfkay` devices too, so maybe something like this (or whatever the proper Python style for that is) could be done:
```suggestion
    if not cmd:
```
(Although if `hid_bootloader_cli` is also not present, then subsequent `cli.run()` will complain about missing `hid_bootloader_cli` even for `halfkay` devices, so maybe a `shutil.which('hid_bootloader_cli')` check would be good too…)

Reply:
The code now checks if `hid_bootloader_cli` is available and use it as a fallback for HalfKay.